MdeModulePkg/Core/Dxe/Image/Image.c

```diff
@@ -663,7 +663,16 @@ CoreLoadPeImage (
   // Or, if our memory protection policy specifies that we shouldn't allow such images, return a failure.
   if (!(Image->ImageContext.SupportsNx) && (Image->ImageContext.ImageType == EFI_IMAGE_SUBSYSTEM_EFI_APPLICATION)) {
     if (gDxeMps.ImageProtectionPolicy.Fields.BlockImagesWithoutNxFlag) {
-      return EFI_SECURITY_VIOLATION;
+      DEBUG ((DEBUG_ERROR, "\n"));
+      DEBUG ((DEBUG_ERROR, "The platform attempted to load an EFI application which does not have the NX_COMPAT DLL Charactersistic.\n"));
+      DEBUG ((DEBUG_ERROR, "This is not allowed by the platform memory protection policy and so the image will not be loaded.\n"));
+      DEBUG ((DEBUG_ERROR, "This can be fixed by setting the NX_COMPAT flag in the image's PE header after validating that the\n"));
+      DEBUG ((DEBUG_ERROR, "image is safe to run with enhanced memory protections enabled. The image can also be executed in compatibility\n"));

```

Comment:
I wonder if we should add a comment here that says where to find the definition of enhanced memory protections. I can easily imagine someone running a platform, seeing this message and going huh, I don't know what enhanced memory protections are, I guess I'll just set compatibility mode. I don't think we should print a web link in the FW, but it is easy to imagine the platform developer looking at the codebase to find this string and coming across the comment.

Reply:
Agreed. I'm adding a section on enhanced memory protection to our mu github.io page so I'll put this on hold til that's complete

Reply:
Added a link to the enhanced memory protection section of the project mu docs